pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,592 @@
+	if err != nil {
+		return err
+	}
+	ipString := ip.String()
+
+	sCopy.Status = kapi.ServiceStatus{
+		LoadBalancer: kapi.LoadBalancerStatus{
+			Ingress: []kapi.LoadBalancerIngress{
+				{
+					IP: ipString,
+				},
+			},
+		},
+	}
+	err = ic.persistStatus(sCopy)
+	if err != nil {
+		ic.release(&ip, key)
+		return err
+	}
+	ic.allocationMap[ipString] = key
+
+	return ic.ensureExternalIP(sCopy, ipString)
+}
+
+// ensureExternalIP ensures that the provided service has the ingress
+// ip persisted as an external ip.
+func (ic *IngressIPController) ensureExternalIP(service *kapi.Service, ingressIP string) error {
+	// Assume it is safe to modify the service without worrying about changing the local cache
+
+	ipExists := false
+	for _, ip := range service.Spec.ExternalIPs {
+		if ip == ingressIP {
+			ipExists = true
+			break
+		}
+	}
+	if !ipExists {
+		service.Spec.ExternalIPs = append(service.Spec.ExternalIPs, ingressIP)
+		return ic.persistSpec(service)
+	}
+	return nil
+}
+
+// allocateIP attempts to allocate the requested ip, and if that is
+// not possible, allocates the next available address.
+func (ic *IngressIPController) allocateIP(requestedIP string) (net.IP, error) {
+	var ip net.IP
+	if requestedIP != "" {
+		ip = net.ParseIP(requestedIP)
+		if ip != nil {
+			// Valid ip
+			err := ic.ipAllocator.Allocate(ip)
+			if err != nil {
+				// Requested ip cannot be allocated
+				ip = nil
+			}
+		}
+	}
+	if ip == nil {
+		return ic.ipAllocator.AllocateNext()
+	} else {
+		return ip, nil
+	}
+}
+
+// release wraps the allocator's Release method to log any error it returns.
+func (ic *IngressIPController) release(ip *net.IP, key string) {
+	err := ic.ipAllocator.Release(*ip)
+	if err != nil {
+		// Release from contiguous allocator should never return an error, but just in case...
+		glog.Errorf("Error releasing ip %v for service %v: %v", ip.String(), key, err)
+	}
+}
+
+// deallocate ensures that the current allocation of a service is
+// removed and that its loadbalancer status is cleared.
+func (ic *IngressIPController) deallocate(service *kapi.Service, key string) error {
+	if !ic.syncProcessed {
+		ic.pendingAllocations = append(ic.pendingAllocations, service)
+		return nil
+	}
+
+	// Make a copy to modify to avoid mutating cache state
+	t, err := kapi.Scheme.DeepCopy(service)
+	if err != nil {
+		return err
+	}
+	sCopy := t.(*kapi.Service)
+
+	err = ic.clearPersistedAllocation(sCopy)
+	if err != nil {
+		return err
+	}
+
+	ipString := sCopy.Status.LoadBalancer.Ingress[0].IP
+	ic.clearLocalAllocation(key, ipString)
+	return nil
+}
+
+func (ic *IngressIPController) persistSpec(service *kapi.Service) error {
+	return ic.persistUpdate(service, false)
+}
+
+func (ic *IngressIPController) persistStatus(service *kapi.Service) error {
+	return ic.persistUpdate(service, true)
+}
+
+func (ic *IngressIPController) persistUpdate(service *kapi.Service, targetStatus bool) error {
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
```

Comment:
Actually, this is the inverse.  So you should be using `wait.ExponentialBackoff` and handling your own exit.